net/client/receivedMessageReader.go

```diff
@@ -0,0 +1,84 @@
+package client
+
+import (
+	"sync"
+
+	"github.com/plgd-dev/go-coap/v3/message/pool"
+	"go.uber.org/atomic"
+)
+
+type ReceivedMessageReaderClient interface {
+	Done() <-chan struct{}
+	ProcessReceivedMessage(req *pool.Message)
+}
+
+type ReceivedMessageReader[C ReceivedMessageReaderClient] struct {
+	queue chan *pool.Message
+	cc    C
+
+	private struct {
+		mutex           sync.Mutex
+		loopDone        chan struct{}
+		readingMessages *atomic.Bool
+	}
+}
+
+// NewReceivedMessageReader creates a new ReceivedMessageReader[C] instance.
+func NewReceivedMessageReader[C ReceivedMessageReaderClient](cc C, queueSize int) *ReceivedMessageReader[C] {
+	r := ReceivedMessageReader[C]{
+		queue: make(chan *pool.Message, queueSize),
+		cc:    cc,
+		private: struct {
+			mutex           sync.Mutex
+			loopDone        chan struct{}
+			readingMessages *atomic.Bool
+		}{
+			loopDone:        make(chan struct{}),
+			readingMessages: atomic.NewBool(true),
+		},
+	}
+
+	go r.loop(r.private.loopDone, r.private.readingMessages)
+	return &r
+}
+
+// C returns the channel to push received messages to.
+func (r *ReceivedMessageReader[C]) C() chan<- *pool.Message {
+	return r.queue
+}
+
+func (r *ReceivedMessageReader[C]) loop(loopDone chan struct{}, readingMessages *atomic.Bool) {
+	for {
+		select {
+		// if the loop is replaced, the old loop will be closed
+		case <-loopDone:
+			return
+		// process received message until the queue is empty
+		case req := <-r.queue:
+			readingMessages.Store(false)
```

Comment:
This code is bit hard to understand:

`readingMessages.Store(true)` is locked by mutex, but this line is not

Reply:
added more comments

Reply:
Yes, this is much better!